### PR TITLE
[WIP] Convergence Criteria

### PR DESCRIPTION
**Pull request by pavanramkumar**
The convergence criteria are now updated to use the norm of the gradient (to be precise, the norm of the update term relative to the norm of the parameters). 

Also, the I found that the issue if fast convergence was not due to a liberal `tol` but due to incorrect initialization of the `glm` object used in cross-validation. Warm restarts were happening inadvertently between CV folds. This is now fixed.

I was able to test this fix with a limited set of simulations but I'd like to test more extensively. This will also likely change the time performance benchmarks.
 
Addresses #230 and #232.

### CODE REVIEW

File: pyglmnet/pyglmnet.py [hunk 1]
```diff
@@ -3,6 +3,7 @@
 from copy import deepcopy
 
 import numpy as np
+import numpy.linalg as ln
```
Comment (jasmainak) [inline pyglmnet/pyglmnet.py]: This is a bit nonstandard abbrev. I would just use the full name linalg

File: pyglmnet/pyglmnet.py [hunk 2]
```diff
@@ -636,19 +628,30 @@ def fit(self, X, y):
...
+        # Iterative updates
         for t in range(0, self.max_iter):
             if self.solver == 'batch-gradient':
                 grad = _grad_L2loss(self.distr,
                                     alpha, self.Tau,
                                     reg_lambda, X, y, self.eta,
                                     beta)
-
+                if t > 1:
+                    if ln.norm(grad) / ln.norm(beta) < tol / lr:
```
Comment (jasmainak) [inline pyglmnet/pyglmnet.py]: isn't this documented somewhere in the docstring?

File: pyglmnet/pyglmnet.py [hunk 3]
```diff
@@ -636,19 +628,30 @@ def fit(self, X, y):
...
+                        break
                 beta = beta - self.learning_rate * grad
+
             elif self.solver == 'cdfast':
+                beta_old = deepcopy(beta)
                 beta, z = \
                     self._cdfast(X, y, z, ActiveSet, beta, reg_lambda)
+                if t > 1:
+                    # pdb.set_trace()
+                    if ln.norm(beta - beta_old) / ln.norm(beta_old) < tol:
```
Comment (jasmainak) [inline pyglmnet/pyglmnet.py]: why do we use different stopping criteria for the two solvers?

**Comment by jasmainak** [comment]
@pavanramkumar can you add a test?

File: pyglmnet/pyglmnet.py [hunk 4]
```diff
@@ -997,22 +986,22 @@ def fit(self, X, y):
...
+                      Tau=self.Tau,
+                      reg_lambda=0.1,
+                      solver=self.solver,
+                      learning_rate=self.learning_rate,
+                      max_iter=self.max_iter,
+                      tol=self.tol,
+                      eta=self.eta,
+                      score_metric=self.score_metric,
+                      random_state=self.random_state,
+                      verbose=self.verbose)
```
Comment (jasmainak) [inline pyglmnet/pyglmnet.py]: so no more warm start?

File: pyglmnet/pyglmnet.py [hunk 5]
```diff
@@ -997,22 +986,22 @@ def fit(self, X, y):
...
+
+            logger.info('Lambda: %6.4f' % rl)
             glm.reg_lambda = rl
```
Comment (jasmainak) [inline pyglmnet/pyglmnet.py]: if we were to follow sklearn convention, you would have a parameter called `warm_start` which can be passed to `GLM` and that would avoid flushing the solutions between multiple fits. And then you would use `set_params` to set the `reg_lambda`

File: pyglmnet/pyglmnet.py [hunk 6]
```diff
@@ -658,20 +661,6 @@ def fit(self, X, y):
                 ActiveSet[beta == 0] = 0
                 ActiveSet[0] = 1.
 
-            # Compute and save loss
-            L.append(_loss(self.distr, alpha, self.Tau, reg_lambda,
```
Comment (jasmainak) [inline pyglmnet/pyglmnet.py]: you also need to remove the `list()` at the top of the loop.

**Comment by pavanramkumar** [comment]
Closes #230

**Comment by jasmainak** [comment]
@pavanramkumar I think we should add a param called `debug` (not documented) which if `True` will trigger the computation of `loss` ...

**Comment by jasmainak** [comment]
superseded by #252
